pool/group.go

```diff
@@ -171,26 +172,46 @@ func (r *Group) Start(ctx context.Context) error {
 		r.wg.Add(1)
 		name := fmt.Sprintf("worker-%d", w)
 
-		go func(name string) {
-			grip.Debugf("worker (%s) waiting for jobs", name)
-		workLoop:
-			for unit := range work {
-				select {
-				case <-ctx.Done():
-					break workLoop
-				default:
-					unit.j.Run()
-					unit.q.Complete(ctx, unit.j)
-				}
-			}
-			grip.Debugf("worker (%s) exiting", name)
-			r.wg.Done()
-		}(name)
+		go groupWorker(ctx, &r.wg, name, work)
 	}
 
 	return nil
 }
 
+func groupWorker(ctx context.Context, wg *sync.WaitGroup, name string, work <-chan workUnit) {
+	var (
+		err  error
+		unit workUnit
+	)
+
+	defer wg.Done()
+	defer func() {
+		// if we hit a panic we want to add an error to the job;
+		err = recovery.HandlePanicWithError(recover(), nil, "worker process encountered error")
+		if err != nil {
+			if unit.j != nil {
+				unit.j.AddError(err)
+				unit.q.Complete(ctx, unit.j)
+			}
+			// start a replacement worker.
+			wg.Add(1)
+			go groupWorker(ctx, wg, name, work)
```

Comment:
Couldn't this continue infinitely, each groupWorker panicking and spawning a new new one?

Reply:
we run this only when there's a panic (e.g. 190)